### ISSUE TITLE
Making sense of *headless-p* and *interactive-p*

### ISSUE THREAD

**Issue by aadcg** [atlas-engineer/nyxt#2695]
Firstly, let's explain why these are different. 

`*headless-p*` defines whether Nyxt is started with a graphical interface, whereas `*interactive-p*` defines whether someone is siting at the machine to send input.

1. **non-headless** and **interactive**. The usual case.
2. **non-headless** and **non-interactive**. The only thing I can imagine in this situation would be starting a graphical Nyxt instance on a remote machine in such a way that it's not possible to send input to it. A rather strange combination.
3. **headless** and **interactive**. Makes sense if the invoking the prompt buffer would fallback on a text-based interface. But we don't have that.
4. **headless** and **non-interactive**. Basically for batch processing.

It should be noted that while the meaning of `*interactive-p*` is generic, we exclusively use it to say whether the prompt buffer can be invoked. 

Notice that 2. is not very interesting, and 3. is not in place. Basically, we care about 1. and 4..

If you invoke the prompt buffer in headless mode, you're greeted with condition `prompt-buffer-non-interactive`. This is rather silly. Why would I want to invoke a graphical prompt buffer on a non-graphical session?

By default, we set `*interactive-p*` to `nil`, and set it to `t` when adequate. For instance, all commands issue via `execute-command` sets it to `t`. 

But then, if you invoke commands such as `watch-mode` on a typical graphical session, the prompt buffer is not invoked! From what I understand, this is because in that lexical environment `*interactive-p*` is set to `t` and the condition that is raised confirming whether you want to invoke the prompt buffer is lost.  This is way we frequently let-bind `*interactive-p*` to `t` when invoking the prompt buffer.  But this is also quite silly. Basically we just want to say that this is an operation that requires a graphical interface so it should raise an error, or fallback on something sensible.

It seems to me that there are too many inconsistencies in the present implementation. Also we're trying to achieve something in a vacuum, since the use cases are extremely marginal.

I suggest getting rid of `*interactive-p*` altogether. That way we can comfortable operate under cases 1. and 4..

**Comment by aartaka** [comment]
Additional context: #2692, #2470.

**Comment by aartaka** [comment]
> 3. **headless** and **interactive**. Makes sense if the invoking the prompt buffer would fallback on a text-based interface. But we don't have that.

Here's when semantics gets funny: `*interactive-p*` in this case means that there's someone inputting and interacting with the program. But it _does not have to be a human being_, it can be init script for headless mode. That's why there's no text interface—we don't need one. We have `nyxt:set-prompt-buffer-input` and that's enough for a script instrumenting Nyxt as an interactive program. And that's the only use-case for both `*headless-p*` and `*interactive-p*` being T—headless automated browser with user-ish input.

> Notice that 2. is not very interesting, and 3. is not in place. Basically, we care about 1. and 4..

3 is in place (,,#ﾟДﾟ)

> If you invoke the prompt buffer in headless mode, you're greeted with condition `prompt-buffer-non-interactive`. This is rather silly. Why would I want to invoke a graphical prompt buffer on a non-graphical session?

See #2470.

> By default, we set `*interactive-p*` to `nil`, and set it to `t` when adequate. For instance, all commands issue via `execute-command` sets it to `t`.
> 
> But then, if you invoke commands such as `watch-mode` on a typical graphical session, the prompt buffer is not invoked! From what I understand, this is because in that lexical environment `*interactive-p*` is set to `t` and the condition that is raised confirming whether you want to invoke the prompt buffer is lost. This is way we frequently let-bind `*interactive-p*` to `t` when invoking the prompt buffer. But this is also quite silly. Basically we just want to say that this is an operation that requires a graphical interface so it should raise an error, or fallback on something sensible.

No. `*interactive-p*` /= "graphical interface". `*interactive-p*` means someone interactng with the program, be it human or another program. No more than that.

> It seems to me that there are too many inconsistencies in the present implementation. Also we're trying to achieve something in a vacuum, since the use cases are extremely marginal.

Use-cases are not marginal, and they were extremely dangerous race-condition-wise before the introduction of `*interactive-p*`, because invoking prompt (which means: creating a prompt buffer view and populating it with content) on wrong thread (like WebKit main thread) is likely to hang the whole browser or crash it. That's why we have to be extremely careful about invoking prompt wherever approptiate and only there. It's not only about graphical/non-graphical, or user/program, or input/no-input distinction, it's also to guard against broken browser.

> I suggest getting rid of `*interactive-p*` altogether. That way we can comfortable operate under cases 1. and 4..

Case 3 is existent and pretty useful for e.g. testing and web scraping.

**Comment by aadcg** [comment]
> Here's when semantics gets funny: *interactive-p* in this case means that there's someone inputting and interacting with the program. But it does not have to be a human being, it can be init script for headless mode. That's why there's no text interface—we don't need one. We have nyxt:set-prompt-buffer-input and that's enough for a script instrumenting Nyxt as an interactive program. And that's the only use-case for both *headless-p* and *interactive-p* being T—headless automated browser with user-ish input.

This has been clear for me from the start. The problem is that, in my understanding, our implementation doesn't reflect these ideas.

> That's why there's no text interface—we don't need one.

Yes, absolutely. Calls to `prompt` are valid both in headless or non-headless modes. In non-headless mode, the user can easily return from it. In headless mode, it is still possible to interact with it but the interaction is done programmatically (not through the GUI). 

When calling `prompt` in headless mode, there should be a warning in the vein of "Behold, you have to interact with this interface programmatically because there are no graphics!".

In conclusion, why do we need `*interactive-p*`? 

. , calls to interactive prompts that require user input in `headless` mode don't make sense! I didn't say that the prompt buffer _can't_ be used in `headless` mode (it clearly can!).

**Comment by aadcg** [comment]
I will not look into #2470 until I extract solid conclusions from this thread.

I have already understood new things and changed my opinion. Probably not for the last time :)

**Comment by aartaka** [comment]
> This has been clear for me from the start. The problem is that, in my understanding, our implementation doesn't reflect these ideas.

How exactly does it not?

> When calling `prompt` in headless mode, there should be a warning in the vein of "Behold, you have to interact with this interface programmatically because there are no graphics!".

Could be useful, yes.

> In conclusion, is our practical use of `*interactive-p*` meaningful (the way it stands now in the codebase)? I see the need for the abstract notion of `*interactive-p*`, but here I'm referring to how we use it.

Again, what exactly seems off to you?

**Comment by aadcg** [comment]
From my top post, I've hinted at concrete issues:

> if you invoke commands such as watch-mode on a typical graphical session, the prompt buffer is not invoked!

**Comment by Ambrevar** [comment]
Repeating myself lest my other comment gets lost: Why not turn `headless-p` into a browser slot?
This would work across all threads.  Slot should be read-only I suppose.

**Comment by Ambrevar** [comment]
I mentioned it in a "commit discussion", but I don't know how GitHub can show me this :p 

Agreed with the start test suite, that's essentially what the current integration tests where aiming to do, but poorly and never reliably.

**Comment by aartaka** [comment]
> I mentioned it in a "commit discussion", but I don't know how GitHub can show me this :p

https://github.com/atlas-engineer/nyxt/commit/bf8f24bf070cba18175fa482217d37698dcb454a#r92833814

**Comment by aadcg** [comment]
> Agreed with the start test suite, that's essentially what the current integration tests where aiming to do, but poorly and never reliably.

Indeed. That goal was too ambitious. I believe that the right target is to test `start.lisp` within the context of the `nyxt` system (not renderer packages). In particular, it is very important to the following check:

- `(nyxt:start :headless t)` => ensure that the `headless` variable (or slot) is rightly set
- Don't clean Nyxt's state
- `(nyxt:start)` => ensure that `headless` fallbacks to the default value.

Obviously, that check must be done for all start parameters.

**Comment by aadcg** [comment]
Closed via #3452. 

The design behind handling `*interactive-p*` was poor and misaligned with the current goals. The idea should be revisited in the future, when Nyxt gets more mature. A proper architecture must be devised.
